### PR TITLE
cmd/clone: add the clone subcommand

### PR DESCRIPTION
**Pull request by zhijian-pro**
Basic functionality is ok, but it hasn't been tested in detail.
- [x] hard link
- [x] add unit test
- [x] clean up the code
- [x] concurrent clone
- [x] progress bar
- [x] sql
- [x] tikv
- [x] check quota
- [x] update dirstat

### CODE REVIEW

File: pkg/meta/redis.go [hunk 1]
```diff
@@ -3755,3 +3756,350 @@ func (m *redisMeta) LoadMeta(r io.Reader) (err error) {
...
+			return m.mkNodeWithAttr(ctx, tx, srcIno, srcAttr, dstParentIno, dstName, dstIno, attach)
+		}, m.inodeKey(srcIno), m.xattrKey(srcIno)); err != nil {
+			return errno(err)
+		}
+		var countDir uint32
+		var maxConcurrency = 4
+		var errCh = make(chan syscall.Errno, maxConcurrency)
+		var wg sync.WaitGroup
+		step := int(math.Ceil(float64(len(entries)) / float64(maxConcurrency)))
+		for i := 0; i < maxConcurrency; i++ {
```
[the remark this answers is not in the document]
Comment (zhijian-pro) [reply pkg/meta/redis.go:3859]: We will adopt the first option.

File: cmd/clone_unix.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+func cmdClone() *cli.Command {
+	return &cli.Command{
+		Name:   "clone",
+		Action: clone,
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:  "cp",
```
[the remark this answers is not in the document]
Comment (SandyXSD) [reply cmd/clone_unix.go:40]:
```
 -p    Cause cp to preserve the following attributes of each source file in the copy: modification time, access time, file flags, file
           mode, user ID, and group ID, as allowed by permissions.  Access Control Lists (ACLs) and Extended Attributes (EAs), including
           resource forks, will also be preserved.
```

[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/clone_unix.go:40]: In my version:

```log
 -p                           same as --preserve=mode,ownership,timestamps
      --preserve[=ATTR_LIST]   preserve the specified attributes (default:
                                 mode,ownership,timestamps), if possible
                                 additional attributes: context, links, xattr,
                                 all
```

File: pkg/meta/tkv.go [hunk 3]
```diff
@@ -3012,3 +3018,278 @@ func (m *kvMeta) LoadMeta(r io.Reader) error {
...
+						return true
+					})
+
+				for indx := uint32(0); indx <= uint32(srcAttr.Length/ChunkSize); indx++ {
+					if v, ok := vals[string(m.chunkKey(srcIno, indx))]; ok {
+						ss := readSliceBuf(v)
+						tx.set(m.chunkKey(*dstIno, indx), v)
+						for _, s := range ss {
+							if s.id > 0 {
+								tx.incrBy(m.sliceKey(s.id, s.size), 1)
```
Comment (davies) [inline pkg/meta/tkv.go:3209]: update them one by one will be slow

File: pkg/meta/tkv.go [hunk 4]
```diff
@@ -3012,3 +3018,278 @@ func (m *kvMeta) LoadMeta(r io.Reader) error {
...
+			if eno := m.doGetAttr(ctx, srcIno, &srcAttr); eno != 0 {
+				return eno
+			}
+			if err := m.mkNodeWithAttr(ctx, tx, srcIno, srcAttr, dstParentIno, dstName, dstIno, cmode, cumask, true); err != nil {
+				return err
+			}
+			// copy chunks
+			if srcAttr.Length != 0 {
+				vals := make(map[string][]byte)
+				tx.scan(m.chunkKey(srcIno, 0), m.chunkKey(srcIno, uint32(srcAttr.Length/ChunkSize)+1),
```
Comment (davies) [inline pkg/meta/tkv.go:3197]: the keys are not ordered by indx

File: pkg/meta/tkv.go [hunk 5]
```diff
@@ -3012,3 +3018,278 @@ func (m *kvMeta) LoadMeta(r io.Reader) error {
...
+		if _, foundIno := m.parseEntry(buf); foundIno != 0 {
+			return syscall.EEXIST
+		}
+	}
+
+	tx.set(m.inodeKey(*dstIno), m.marshal(&srcAttr))
+
+	// copy xattr
+	tx.scan(m.xattrKey(srcIno, ""), nextKey(m.xattrKey(srcIno, "")), false, func(k, v []byte) bool {
+		tx.set(m.xattrKey(*dstIno, string(k)), v)
```
Comment (davies) [inline pkg/meta/tkv.go:3286]: `k` has the prefix

File: pkg/meta/base_test.go [hunk 6]
```diff
@@ -2019,3 +2021,326 @@ func testDirStat(t *testing.T, m Meta) {
...
+	sort.Slice(entries1, func(i, j int) bool { return string(entries1[i].Name) < string(entries1[j].Name) })
+	sort.Slice(entries2, func(i, j int) bool { return string(entries2[i].Name) < string(entries2[j].Name) })
+	for idx, entry := range entries1 {
+		if string(entry.Name) == "." || string(entry.Name) == ".." {
+			continue
+		}
+		if entry.Attr.Typ == TypeDirectory {
+			checkEntryTree(t, m, entry.Inode, entries2[idx].Inode, walkFunc)
+		}
+		walkFunc(entry, entries2[idx], dstIno)
```
Comment (davies) [inline pkg/meta/base_test.go:2322]: check xattr

File: cmd/clone.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	srcIno, err := utils.GetFileInode(srcPath)
+	if err != nil {
+		return fmt.Errorf("lookup inode for %s: %s", srcPath, err)
+	}
+	dst := ctx.Args().Get(1)
+	if strings.HasSuffix(dst, "/") {
+		dst = filepath.Join(dst, filepath.Base(srcPath))
+	}
+	if _, err := os.Stat(dst); err == nil || !os.IsNotExist(err) {
+		return fmt.Errorf("%s already exists", dst)
```
Comment (SandyXSD) [inline cmd/clone.go:67]: print the error

File: cmd/clone.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	dstAbsPath, err := filepath.Abs(dst)
+	if err != nil {
+		return fmt.Errorf("abs of %s: %s", dstAbsPath, err)
```
Comment (SandyXSD) [inline cmd/clone.go:71]: the first argument should be `dst`

File: pkg/meta/redis.go [hunk 9]
```diff
@@ -3811,3 +3816,319 @@ func (m *redisMeta) LoadMeta(r io.Reader) (err error) {
...
+				return err
+			}
+			path, err := m.rdb.Get(ctx, m.symKey(srcIno)).Result()
+			if err != nil {
+				return err
+			}
+			return m.rdb.Set(ctx, m.symKey(*dstIno), path, 0).Err()
+		}, m.symKey(srcIno))
+
+	case TypeBlockDev, TypeCharDev, TypeFIFO, TypeSocket:
```
Comment (SandyXSD) [inline pkg/meta/redis.go:4052]: default?

File: pkg/meta/redis.go [hunk 10]
```diff
@@ -3811,3 +3816,319 @@ func (m *redisMeta) LoadMeta(r io.Reader) (err error) {
...
+					return errno(err)
+				}
+			}
+			m.updateParentStat(ctx, *dstIno, srcAttr.Parent, int64(srcAttr.Length), align4K(srcAttr.Length))
+			return nil
+		}, m.inodeKey(srcIno), m.xattrKey(srcIno))
+	case TypeSymlink:
+		err = m.txn(ctx, func(tx *redis.Tx) error {
+			var srcAttr Attr
+			if eno := m.doGetAttr(ctx, srcIno, &srcAttr); eno != 0 {
```
Comment (SandyXSD) [inline pkg/meta/redis.go:4039]: use `tx`

File: pkg/meta/redis.go [hunk 11]
```diff
@@ -3811,3 +3816,319 @@ func (m *redisMeta) LoadMeta(r io.Reader) (err error) {
...
+				return eno
+			}
+			if err := m.mkNodeWithAttr(ctx, tx, srcIno, &srcAttr, dstParentIno, dstName, dstIno, cmode, cumask, true); err != nil {
+				return err
+			}
+			path, err := m.rdb.Get(ctx, m.symKey(srcIno)).Result()
```
Comment (SandyXSD) [inline pkg/meta/redis.go:4045]: use `tx`

File: pkg/meta/redis.go [hunk 12]
```diff
@@ -3811,3 +3816,319 @@ func (m *redisMeta) LoadMeta(r io.Reader) (err error) {
...
+			if ctx.Canceled() {
+				return syscall.EINTR
+			}
+			entries[i] = nil // release memory
+		}
+		wg.Wait()
+		// reset nlink
+		if attach && countDir+2 != srcNlink {
+			var dstAttr Attr
+			if eno := m.doGetAttr(ctx, *dstIno, &dstAttr); eno != 0 {
```
Comment (SandyXSD) [inline pkg/meta/redis.go:3985]: we already have the attr at the beginning

**Pull request merged**